internal/commands/build/build.go

```diff
@@ -0,0 +1,240 @@
+	flags.StringVar(&opts.progress, "progress", "auto", "Set type of progress output (auto, plain, tty). Use plain to show container output")
+	flags.BoolVar(&opts.pull, "pull", false, "Always attempt to pull a newer version of the image")
+	flags.StringVarP(&opts.out, "output", "o", "", "Dump generated bundle into a file")
+	flags.StringVarP(&opts.tag, "tag", "t", "", "Name and optionally a tag in the 'name:tag' format")
+
+	return cmd
+}
+
+func runBuild(dockerCli command.Cli, application string, opt buildOptions) (reference.Named, error) {
+	err := checkMinimalEngineVersion(dockerCli)
+	if err != nil {
+		return nil, err
+	}
+
+	var ref reference.Named
+	ref, err = packager.GetNamedTagged(opt.tag)
+	if err != nil {
+		return nil, err
+	}
+
+	app, err := packager.Extract(application)
+	if err != nil {
+		return nil, err
+	}
+	defer app.Cleanup()
+
+	bundle, err := packager.MakeBundleFromApp(dockerCli, app, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	buildopts, err := parseCompose(app, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	buildopts["invocation-image"], err = createInvocationImageBuildOptions(dockerCli, app)
+	if err != nil {
+		return nil, err
+	}
+
+	debugBuildOpts(buildopts)
+
+	ctx, cancel := context.WithCancel(appcontext.Context())
+	defer cancel()
+	d, err := driver.GetDriver(ctx, "buildx_buildkit_default", nil, dockerCli.Client(), nil, "", nil)
+	if err != nil {
+		return nil, err
+	}
+	driverInfo := []build.DriverInfo{
+		{
+			Name:   "default",
+			Driver: d,
+		},
+	}
+
+	pw := progress.NewPrinter(ctx, os.Stderr, opt.progress)
+
+	// We rely on buildx "docker" builder integrated in docker engine, so don't nee a DockerAPI here
+	resp, err := build.Build(ctx, driverInfo, buildopts, nil, dockerCli.ConfigFile(), pw)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println("Successfully built service images")
+	updateBundle(bundle, resp)
+
+	if ref == nil {
+		if ref, err = computeDigest(bundle); err != nil {
+			return nil, err
+		}
+	}
+
+	if err = persistBundle(opt, bundle, ref); err != nil {
+		return nil, err
+	}
+	return ref, nil
+}
+
+func checkMinimalEngineVersion(dockerCli command.Cli) error {
+	info, err := dockerCli.Client().Info(appcontext.Context())
+	if err != nil {
+		return err
+	}
+	majorVersion, err := strconv.Atoi(info.ServerVersion[:strings.IndexRune(info.ServerVersion, '.')])
+	if err != nil {
+		return err
+	}
+	if majorVersion < 19 {
+		return errors.New("'build' require docker engine 19.03 or later")
+	}
+	return nil
+}
+
+func updateBundle(bundle *bundle.Bundle, resp map[string]*client.SolveResponse) {
+	debugSolveResponses(resp)
+	for service, r := range resp {
+		digest := r.ExporterResponse["containerimage.digest"]
+		if service == "invocation-image" {
+			bundle.InvocationImages[0].Digest = digest
+		} else {
+			image := bundle.Images[service]
+			image.Image = fmt.Sprintf("%s:%s-%s", bundle.Name, bundle.Version, service)
```

Comment:
I don't understand this one. Are you always overriding the image name with something in the form `name:version-service` ? Shouldn't we do that IFF the image name is not set?

Reply:
No specific opinion on tag to use here. I assume once used in a bundle an image is only identified by content-digest, then repo@digest once pushed to a registry

Reply:
I don't think so, the bundle should reflect what is inside the compose file (as we do now). The content digests will be fixed during the push, directly in the OCI index. The content digests will be injected in the invocation image to patch the compose file "on-the-fly" using the `relocation map` (that's what I'm working on). So as an example, the bundle should refer to "nginx", but the OCI index will refer the digested reference of the mounted nginx image inside the app repository. Then while pulling the app, we will store the relocation mapping `nginx` -> `myapp@sha256:beefc0ffee...` in the bundle store, along with the `bundle.json` itself so we can inject it at runtime.
Does that make sense?